src/renderer/base/renderer.cpp

```diff
@@ -786,7 +786,7 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
             {
                 COORD thisPoint{ screenPoint.X + gsl::narrow<SHORT>(cols), screenPoint.Y };
                 const auto thisPointPatterns = _pData->GetPatternId(thisPoint);
-                if (color != it->TextAttr() || patternIds != thisPointPatterns)
+                if (color != it->TextAttr() || (thisPointPatterns.size() > 0 && patternIds != thisPointPatterns))

```

Comment:
This change is crucial, and reduces 3% of the CPU time?

How could comparison between two *empty* vectors be so expensive 😓

Reply:
Nope. This is wrong. This caused incorrect rendering.

Reply:
If you want to optimize something, you could fix `Terminal::GetPatternId`.
It's using `findOverlapping()` which allocates a `std::vector<interval_tree::Interval<>>` internally, just to then go ahead and map the vector back to another `std::vector<size_t>`. For every character we draw...
It might "only" be 5% CPU usage in your trace, but it's actually 10%, because it then goes ahead to compare the vector with `patternIds`. This puts an insane pressure on the CPU's MMU, caches, and makes all other allocations slower as well.